### PR TITLE
Delete announcements

### PR DESCRIPTION
**Pull request by wwohlers**

### CODE REVIEW

File: src/components/Modals/AnnouncementModal.vue [hunk 1]
```diff
@@ -44,11 +62,22 @@ export default {
...
+        alert("Error: " + e);
+      }
+    },
   },
 };
 </script>
 
-<style scoped>
+<style lang="less" scoped>
+  @import '../../../assets/global-classes.less';
```
Comment (ZachOBrien) [inline src/components/Modals/AnnouncementModal.vue:80]: nice, thanks for using this

File: src/components/Modals/AnnouncementModal.vue [hunk 2]
```diff
@@ -6,7 +6,13 @@
           <div class="announce-title">{{announcement.title}}</div>
           <div class="announce-date">{{ createdString }}</div>
         </div>
-        <div class="announce-body">{{announcement.description}}</div>
+        <div class="announce-body">
+          {{announcement.description}}
+          <access-control :roles="[USER[ROLE.ADMIN]]">
+            <br/>
+            <button class="btn--warning" @click="deleteAnnouncement">Delete</button>
+          </access-control>
```
Comment (ZachOBrien) [inline src/components/Modals/AnnouncementModal.vue:14]: Good use of the access control component here

File: src/components/Modals/AnnouncementModal.vue [hunk 3]
```diff
@@ -32,6 +41,15 @@ export default {
       type: Object,
     },
   },
+  data() {
+    return {
+      USER,
+      ROLE,
+    };
```
Comment (ZachOBrien) [inline src/components/Modals/AnnouncementModal.vue:48]: Because this data is constant and we don't need it to be reactive, we can just instantiate it in a created() hook. Just in case you want it, an outline of that process is explained at the bottom of [this article](https://coderethinked.com/3-different-ways-to-access-constants-in-a-vue-template/ ). It's a small change but just in the spirit of best practices.

File: src/views/ProfileView.vue [hunk 4]
```diff
@@ -111,6 +114,9 @@ export default {
     closeAnnouncementModal() {
       this.openModal = false;
     },
+    deleteAnnouncement() {
+      this.announcementsListKey += 1; // reloads the announcements-list component
+    },
```
Comment (ZachOBrien) [inline src/views/ProfileView.vue:119]: This is a clever solution for the component as it is. Needing to do this though is a hint that we should refactor the underlying methodology of grabbing announcements out into the store.

File: src/components/Announcements/AnnouncementsList.vue [hunk 5]
```diff
@@ -9,6 +9,7 @@
       <div class="announcement-card">
         <div class="announce-header">
           <div class="announce-title">{{a.title}}</div>
+          <div>{{"id " + a.id}}</div>
```
Comment (justinkonecny) [inline src/components/Announcements/AnnouncementsList.vue:12]: Do we need to show Announcement ID?

File: .gitignore [hunk 6]
```diff
@@ -3,4 +3,4 @@ npm-debug.log
 /dist
 *.xml
 .idea/
-*.DS_Store
+*.DS_Store
```
Comment (jackblanc) [inline .gitignore:6]: Can you undo this change? we want a newline at the end of the file

File: src/components/Modals/AnnouncementModal.vue [hunk 7]
```diff
@@ -6,7 +6,14 @@
           <div class="announce-title">{{announcement.title}}</div>
           <div class="announce-date">{{ createdString }}</div>
         </div>
-        <div class="announce-body">{{announcement.description}}</div>
+        <div class="announce-body">
+          {{announcement.description}}
+          <access-control :roles="[USER[ROLE.ADMIN]]">
+            <br/>
```
Comment (jackblanc) [inline src/components/Modals/AnnouncementModal.vue:12]: Why do we need the `<br />`? We should avoid using those where possible

File: src/components/Modals/AnnouncementModal.vue [hunk 8]
```diff
@@ -30,8 +39,16 @@ export default {
     },
     announcement: {
       type: Object,
+      required: true,
     },
   },
+  created() {
+    this.USER = USER;
```
Comment (jackblanc) [inline src/components/Modals/AnnouncementModal.vue:46]: Why is this necessary? where are these variables referenced?

File: src/store/modules/announcements.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    getEventAnnouncements(state, eventId) {
+      return state.announcements.filter(
+        ann => ('event_id' in ann) && ann.event_id === eventId,
+      );
+    },
+  },
+  mutations: {
+    loadAnnouncements(state, { announcements }) {
+      announcements.forEach((ann) => {
+        if (state.announcements.filter(a => a.id === ann.id).length === 0) {
```
Comment (jackblanc) [inline src/store/modules/announcements.js:23]: This seems a little overcomplicated/inefficient. There should be a contains function we can use?

File: src/components/Modals/AnnouncementModal.vue (hunk 8 above)
Comment (ZachOBrien) [reply src/components/Modals/AnnouncementModal.vue:46]: These are used in the :roles prop for AccessControl. Line 11 in this file.

File: src/components/Modals/AnnouncementModal.vue (hunk 7 above)
Comment (ZachOBrien) [reply src/components/Modals/AnnouncementModal.vue:12]: It appears this was just adding space. I took it out and added some padding with CSS instead.

File: .gitignore (hunk 6 above)
Comment (ZachOBrien) [reply .gitignore:6]: fixed

File: src/store/modules/announcements.js (hunk 9 above)
Comment (ZachOBrien) [reply src/store/modules/announcements.js:23]: I had trouble getting that working. I tried doing this: 
image

but the conditional always returned true and so whenever something new gets loaded, all the old announcements stay and get duplicated. I'm not familiar with the technicalities of how javascript checks for equality. Do we have to override something like equals() and hashCode() in Java?

Comment (jackblanc) [reply src/store/modules/announcements.js:23]: I think we might want to store data in Vuex differently so that we don't need this logic. Storing all announcements in one array requires us to have this logic to create that array, and logic in the getters to filter the array. A better solution might be storing sitewide announcements and event announcements separately.

Comment (ZachOBrien) [reply src/store/modules/announcements.js:23]: Gotcha. I can change that

File: src/store/modules/announcements.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
+import api from '../../api/api';
+
+export default {
+  namespaced: true,
+  state: {
+    announcements: [],
```
Comment (jackblanc) [inline src/store/modules/announcements.js:6]: I'm confused, I thought you said this had been separated out? https://c4cneuteams.slack.com/archives/C016RC79ZD2/p1598574066015400?thread_ts=1598568973.013300&cid=C016RC79ZD2

File: src/views/SingleEventView.vue [hunk 11]
```diff
@@ -168,8 +168,8 @@ export default {
     AnnouncementModal,
   },
   props: {
-    eventId: { // id is a number, but props are always passed as strings
-      type: String,
+    eventId: {
+      type: Number,
       required: true,
```
Comment (ZachOBrien) [inline src/views/SingleEventView.vue:173]: I added a method to the router to turn this prop into a Number so we don't have to mess with parsing later on

File: src/router/router.js [hunk 12]
```diff
@@ -68,7 +68,11 @@ const allRoutes = [
     path: '/event/:eventId',
     name: 'single-event',
     component: SingleEventView,
-    props: true,
+    props(route) {
```
Comment (jackblanc) [inline src/router/router.js:71]: why is this needed?

File: src/store/modules/announcements.js [hunk 13]
```diff
@@ -1,52 +1,84 @@
...
+      } else {
+        state.sitewideAnnouncements = removeAnnouncement(
+          state.sitewideAnnouncements, payload.announcement.id,
+        );
+      }
     },
     clearAnnouncements(state) {
-      state.announcements = [];
+      state.sitewideAnnouncements = [];
+      state.eventSpecificAnnouncements = new Map();
```
Comment (jackblanc) [inline src/store/modules/announcements.js:67]: should this be an array?

**Review by jackblanc — DISMISSED** [review]
I have a few questions but nothing major, good work here getting this together 👍

File: src/router/router.js (hunk 12 above)
Comment (ZachOBrien) [reply src/router/router.js:71]: Without this, the event ID is passed to components not as a number but as a string. That was causing issues where we had to use parseInt() and have eslint ignore that line. This just makes it so we don't have to parseInt.

**Review by jackblanc — APPROVED** [review]
Looks good to me as long as you've verified locally 🔥
